### PR TITLE
WT-9432 Ensure Evergreen builds have debug information

### PR DESCRIPTION
**Pull request by agorrod**
Do this by setting the optimization level to 0

### CODE REVIEW

**Comment by agorrod** [comment]
Hold off on review of this for now. I think I can make a smaller change in the cmake/config files that has a better outcome.

File: cmake/configs/base.cmake
Comment (korteland) [inline cmake/configs/base.cmake:217]: I think this could be a little clearer - we're always setting the default build type, but in this clause we're "Defaulting build type to ..."

File: cmake/configs/base.cmake
Comment (korteland) [inline cmake/configs/base.cmake:226]: Have you considered `-Og` for this? Not sure if MSVC has an equivalent.

Also, we should announce this change in a team channel. I (and I suspect a few other people) usually recycle a bunch of "known good" CMake flags whenever I'm building, so it's a chance to update people's assumptions about optimisation levels. (e.g. assuming optimised local builds could be bad when working on a server-related issue).

File: cmake/configs/modes.cmake
Comment (korteland) [inline cmake/configs/modes.cmake:133]: Nitpick: the comment applies to all sanitiser builds, not just ASAN

File: cmake/configs/modes.cmake
Comment (korteland) [inline cmake/configs/modes.cmake:134]: Same comment about `-Og`, but do we also need to force `-g` here? For some reason I thought the sanitiser builds already used `-g` (via `HAVE_DIAGNOSTIC`) but I could be wrong.

File: cmake/configs/base.cmake
Comment (mickgraham-mongodb) [reply cmake/configs/base.cmake:226]: Instead of /O0 for windows it might be /Od

https://docs.microsoft.com/en-us/cpp/build/reference/od-disable-debug?view=msvc-170

Comment (agorrod) [reply cmake/configs/base.cmake:226]: I did look at `Og`, the clang documentation says:
```[-Og] Like [-O1]. In future versions, this option might disable different optimizations in order to improve debuggability.```

I'm inclined to go with the most extreme position here, but I could be convinced otherwise. I'd originally thought we were using the default of O1 here, but since it's O3 at the moment, O1 might be OK?

Ack re: Od - makes sense.

File: cmake/configs/modes.cmake
Comment (agorrod) [reply cmake/configs/modes.cmake:133]: Thanks - fixed.

File: cmake/configs/modes.cmake
Comment (agorrod) [reply cmake/configs/modes.cmake:134]: I agree that the HAVE_DIAGNOSTIC code turns `-g` on for some cases, including this at the moment. It seems to me like a tenuous connection to link a compile time define in WiredTiger with build flags.

I'm not sure how big to make this change, so I didn't want to go messing with that code right here. I'd be OK to expand the scope, but I think that would require a more careful review of evergreen.yml and it's usage of HAVE_DIAGNOSTIC to make sure we got things right.

FWIW: I included this here to (hopefully) match the default cmake_build_type switch as well. I'm presuming that debug builds will include disabling optimizations and outputting debug information.

File: cmake/configs/base.cmake
Comment (korteland) [reply cmake/configs/base.cmake:230]: I did a little experimenting, -O1 doesn't seem to optimise local variables away or inline stuff. Sorry to press the matter but I'd still prefer -O1 - completely turning off all optimisations seems like a bit too big of a hammer. (In my experience, -O1 will get you about 70% of the benefits of -O3 anyway.)

File: cmake/configs/modes.cmake
Comment (korteland) [reply cmake/configs/modes.cmake:134]: > It seems to me like a tenuous connection to link a compile time define in WiredTiger with build flags.

That's fair. If the goal is to guarantee a particular optimisation level and debugging flags for sanitiser builds no matter what, then this is the way to do it. One thing to be aware of here is that these flags won't be override-able from the command line since they're the final occurrence of these options - not sure how much we care about that for sanitiser builds.

Comment (agorrod) [reply cmake/configs/modes.cmake:134]: I considered making explicit release and debug versions of these build types, but figured it wasn't necessary at the moment. I'm not convinced that overriding the build_type is the right approach to Sanitizer builds moving forward, so I'm inclined to take this, and resolve the generalization in the future when we tidy up the build/test setup more broadly.

File: cmake/configs/base.cmake
Comment (agorrod) [reply cmake/configs/base.cmake:230]: I went and looked at the GCC docs, and they make a compelling case for -Og - I'll go ahead and make the change.

> -Og should be the optimization level of choice for the standard edit-compile-debug cycle, offering a reasonable level of optimization while maintaining fast compilation and a good debugging experience. It is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0.

**Review by mickgraham-mongodb — APPROVED** [review]
LGTM

**Review by korteland — APPROVED** [review]
Thanks, LGTM now!

**Comment by agorrod** [comment]
FWIW: The test failure is there in the baseline as well.

@korteland would you mind taking yet another look before I go ahead and merge?

**Review by korteland — APPROVED** [review]
Still looks good :-)

**Pull request merged**
